File: tests/test_lottery.py
# Tests for Lottery yourToken — run: python tests/test_lottery.py
import math
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

from lottery_yourtoken import (  # noqa: E402
    create_lottery, decode, difficulty, dumps, encode, generate_key,
    generate_password, is_permutation, loads, reveal_key, shuffle_password,
    verify_guess,
)

PASSED = 0


def ok(cond, name):
    global PASSED
    assert cond, name
    PASSED += 1
    print(f"  ok  {name}")


# Fast KDF for tests (the construction is iteration-count agnostic).
FAST = {"iterations": 1_000}

KEY = generate_key(64)
BUNDLE, PASSWORD = create_lottery(KEY, **FAST)

# --- lifecycle -------------------------------------------------------------
ok(reveal_key(BUNDLE, PASSWORD) == KEY, "correct ordering reveals the exact key")
ok(verify_guess(BUNDLE, PASSWORD), "correct ordering verifies")
ok(BUNDLE["anagram"] != PASSWORD, "published anagram is not the answer")
ok(is_permutation(BUNDLE["anagram"], PASSWORD), "anagram is a permutation of the answer")
ok(len(PASSWORD) == 30, "password has 30 characters")

# --- wrong guesses ---------------------------------------------------------
wrong = PASSWORD[1:] + PASSWORD[0]
if wrong == PASSWORD:
    wrong = PASSWORD[::-1]
ok(not verify_guess(BUNDLE, wrong), "wrong ordering fails verification")
ok(not verify_guess(BUNDLE, "x" * 30), "non-permutation is rejected")
ok(not verify_guess(BUNDLE, PASSWORD[:-1]), "shorter candidate is rejected")
try:
    reveal_key(BUNDLE, wrong)
    ok(False, "reveal_key must raise on a losing candidate")
except ValueError:
    ok(True, "reveal_key raises on a losing candidate")

# --- tamper detection ------------------------------------------------------
import copy  # noqa: E402

tampered = copy.deepcopy(BUNDLE)
ct = bytes.fromhex(tampered["cipher"]["ciphertext"])
tampered["cipher"]["ciphertext"] = bytes([ct[0] ^ 1]) .hex() + ct[1:].hex()
ok(not verify_guess(tampered, PASSWORD), "tampered ciphertext fails the MAC")

# --- determinism -----------------------------------------------------------
salt = bytes(range(16))
b1, _ = create_lottery("secret-key", password="abcabcabcabcabcabcabcabcabcabc",
                       salt=salt, **FAST)
b2, _ = create_lottery("secret-key", password="abcabcabcabcabcabcabcabcabcabc",
                       salt=salt, **FAST)
ok(b1["cipher"]["ciphertext"] == b2["cipher"]["ciphertext"]
   and b1["mac"]["tag"] == b2["mac"]["tag"],
   "same password + salt => same ciphertext and tag (auditable)")

# --- difficulty ------------------------------------------------------------
ok(difficulty("aab")[0] == 3, "difficulty('aab') = 3 orderings")
n30, bits30 = difficulty("abcdefghjkmnpqrstuvwxyz2345678")  # 30 unique chars
ok(n30 == math.factorial(30), "30 unique characters => 30! orderings")
ok(abs(bits30 - math.log2(math.factorial(30))) < 1e-9, "log2 difficulty matches")
full, _ = difficulty(BUNDLE["anagram"])
hinted, _ = difficulty(BUNDLE["anagram"], hints=5)
ok(hinted < full, "hints reduce the search space")
ok(difficulty("abc", hints=3)[0] == 1, "all positions hinted => 1 ordering")

# --- serialization ---------------------------------------------------------
ok(loads(dumps(BUNDLE))["mac"]["tag"] == BUNDLE["mac"]["tag"],
   "bundle JSON round-trips")

# --- Blockz10 {e,1} tie-in -------------------------------------------------
ok(set(KEY) <= {"e", "1"} and len(KEY) == 64, "prize key is a 64-char {e,1} string")
ok(decode(encode(KEY)) == KEY, "Blockz10 encoding round-trips the prize key")
ok(encode("eee11") == "311", "blog example: eee11 -> 311")

# --- shuffle sanity ---------------------------------------------------------
p = generate_password()
ok(sorted(shuffle_password(p)) == sorted(p), "shuffle preserves the multiset")

print(f"\n{PASSED} tests passed.")

File: tests/test_lottery.mjs
// Cross-language test: the JS mirror must accept the Python-made vector.
// Run: node tests/test_lottery.mjs
import { readFileSync } from "node:fs";
import {
  bzDecode, bzEncode, createLottery, difficulty, isPermutation,
  revealKey, shufflePassword, verifyGuess,
} from "../src/js/lottery.js";

let passed = 0;
const ok = (cond, name) => {
  if (!cond) { console.error(`FAIL  ${name}`); process.exit(1); }
  passed++; console.log(`  ok  ${name}`);
};

const vector = JSON.parse(readFileSync(new URL("./vector.json", import.meta.url)));
const PASSWORD = "abcdefghjkmnpqrstuvwxyz2345678";

ok(await verifyGuess(vector, PASSWORD), "JS verifies the Python-made bundle");
ok((await revealKey(vector, PASSWORD)) === "e1".repeat(32),
  "JS decrypts the Python ciphertext to the exact key");
ok(!(await verifyGuess(vector, PASSWORD.split("").reverse().join(""))),
  "JS rejects a wrong ordering of the same characters");

// Native JS round-trip.
const { bundle, password } = await createLottery("11ee11ee", null, { iterations: 1000 });
ok(await verifyGuess(bundle, password), "JS round-trip verifies");
ok((await revealKey(bundle, password)) === "11ee11ee", "JS round-trip decrypts");
ok(bundle.anagram !== password, "JS anagram is not the answer");
ok(isPermutation(bundle.anagram, password), "JS anagram is a permutation");

// Difficulty + encoding parity with Python.
ok(difficulty("aab").orderings === 3n, "difficulty('aab') = 3");
ok(difficulty("abc", 3).orderings === 1n, "all hints => 1 ordering");
ok(bzEncode("eee11") === "311" && bzDecode("311") === "eee11",
  "Blockz10 encoding: eee11 <-> 311");
ok(sortedEq(shufflePassword("abcabc"), "abcabc"), "shuffle preserves multiset");

function sortedEq(a, b) {
  return [...a].sort().join("") === [...b].sort().join("");
}

console.log(`\n${passed} tests passed.`);
